### ISSUE TITLE
Discussion of usage in a Qt application

### ISSUE THREAD

**Issue by axxel** [nu-book/zxing-cpp#128]
As @vkrause pointed out in https://github.com/nu-book/zxing-cpp/issues/86#issuecomment-641393971, there are a couple of KDE apps using this library. I had a look at the three code bases and have a few comments / suggestions:

Two of the apps only deal with decoding QRCodes and share the following issues
 * they set `tryRotate()`, even that has no effect on the QRCode decoder
 * they could set `tryHarder(true)`, which should be considered the default (see also #104)
 * they create a `GenericLuminanceSource` with the RGB constructor overload but set the `pixelBytes` parameter to 1, which is a bug in any case (is this particular QImage guaranteed to contain only grey-scale bits?)
 * they could both take advantage of the 'new' `ReadBarcode` function and simplify the code that way or alternatively go one level deeper and use `ZXing::QRCode::Reader`
 * they could use `QString::fromWCharArray` (see [here](https://stackoverflow.com/questions/14726304/best-way-to-convert-stdwstring-to-qstring)) instead of using the `TextUtfEncoding`

The other deals with scanning several 2D codes. It provides it's own `BinaryImage` implementation as a wrapper for `QImage` but would effectively only work properly with 'ideal' black/white input. Why is that code not using the ZXing binarizer code? Performance?

Also related to the discussion in #75, I'd like to start some trivial Qt support basically offering a `ReadBarcode(const QImage&, ...)` function. This can be implemented in a header-only way (not introducing any Qt build dependency) and prevent most of the issues above.

**Comment by vkrause** [comment]
Thanks, I'll look into applying the points for the QR decoder.

For the custom `BinaryImage` case, yes, this is done for performance, and only needs to work on "ideal" input indeed (barcodes taken from raster or vector data in PDF files or email attachments). This looks for any kind of boarding pass or train ticket, and needs to finish this within a few milliseconds.

Regarding Qt wrappers: Those can indeed be convenient, especially as long as you have a somewhat common use-case. For the itinerary extractor I actually started with qzxing, but the additional image copy that brought in was not worth the minor convenience gain, even before doing the optimizations for "ideal" input, but that is certainly a more niche case.

**Comment by axxel** [comment]
Before you change that code for the QR decoder, would you wait for my go at some trivial QImage support or do you want to have it compatible with 'old' ZXing code anyway?

About the 'ideal' case: if you are only scanning for QRCodes then omitting the ZXing binarizers is definitively worth it. Also setting `tryHarder(true)` makes no sens at all in that case.

Questions:
 1. why did you choose to surround your input with one pixel wide white border?
 1. you mentioned a regression after 1.0.6 in your `QImagePureBinarizer::isPureBarcode()` method but I could not find the hash in the comment? I was wondering what happened there.

**Comment by axxel** [comment]
I've decided to (finally) extend the `ReadBarcode` API to support all your use cases with a single function call without having to resort to the (IMHO) 'over-engineered' `LuminanceSource` interface which is only there because the upstream project had it. I'll not not remove that but hope that code will migrate away from it (sooner than later...)

The problem at hand is really trivial from a user story point of view: You have a 2D image buffer with some internal format and a set of options (`DecodeHints`) and you want to call a single function to get your barcode back. That is what all 'wrappers' and samples ended up doing. I currently fail to see _any_ other use case for the code (apart from the completely separate writing/encoding path, which is equally trivial).

ad (1): Several (actually almost all) codes require a certain quite zone to be valid. But I think it is fair use to be able to provide a 'pure' (meaning generated/perfect) code without any quite zone. I will make sure the new 'direct' support for those will do something similar.

ad (2): thanks for pointing that out. I'll have a look. For the future: please feel free to report those issues when you detect them :). I'd like to add your samples to our own testing code, once that pure support is properly set up and running.

**Comment by axxel** [comment]
Done: 
 * first trivial Qt wrapper in form of an [example application](https://github.com/nu-book/zxing-cpp/blob/master/example/ZXingQtReader.cpp) to have a look at
 * deprecated the `isPureBarcode` function (is never called anymore)
 * fixed the regression that caused @vkrause's pure-code-path to crash at some point in time (i hope)

ToDo:
 * automatic quite zone extension/insertion for the 'pure' barcodes

Feedback welcome. Also from @swex maybe?

**Comment by axxel** [comment]
I just added a new `ZXingQtCamReader` example that is something like a stripped down version of what @swex QZingNu wrapper does. The noticeable difference is that it does away with all the `QVideoFrame` to `QImage` conversion and uses the video frame data directly. That removed two image copies and all pixel conversions from YUV -> RGB -> Grey. The code is simpler, faster and also produces no additional noise during the pixel format conversion. Disclaimer: most of the 'supported' pixel formats from QVideoFrame are not tested, due to lack of input.

There are now overloads for `ReadBarcode` that take a `QImage` or a `QVideoFrame`. Additionally, there is a video filter that can be used in QML applications.

What is still pretty much a work in progress is the right way to export the public ZXing enums, like `BarcodeFormat` without repeating their definition and introducing a second, incompatible enum class.

My intention would be 'officially' include the `ZXingQtReader.h` as a header only access to the wrapper without adding a separate library linked to qt to the project. This will require a little more care with the moc generation when incorporating that code to a client project. I'd appreciate any feedback and help related to best practices regarding the moc/meta-data setup.

**Comment by vkrause** [comment]
With master breaking API we use I was now finally forced to look at this ;-)

Two observations so far:
- Grayscale QImages don't seem to work, e.g. https://invent.kde.org/pim/kitinerary/-/blob/master/autotests/barcodes/pdf417_8bit.png fails for me while the same code in various other formats in the same folder works perfectly fine, and so does that image converted to another format.
- "pure" QR codes without a silent zone don't work, which isn't even technically wrong I think. But since we get such images from PDF files we are adding padding for this in our code. Doing that in a way not needing an extra copy doesn't seem possible anymore, as ImageView assumes continuous image memory?

**Comment by axxel** [comment]
I was not aware that I broke something and it was definitively not intentional but if I got your attentions this way, works for me ;).

Can you tell me what API I broke?

Ad your issues:
 * Grayscale QImage input see https://github.com/nu-book/zxing-cpp/commit/adbd7a2c44dafb4b7d4e74df575b85faf38ca42e
 * "pure" QR Codes without a quite zone do work, see e.g. [11.png](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/qrcode-2/11.png). Can you provide a sample?

**Comment by vkrause** [comment]
The change that broke for us is https://github.com/nu-book/zxing-cpp/commit/c853e1989032b2e915c4022118e32ffab4dd3c27#diff-7b9991de44b2e29ece283e519abf79d03dc98f4f4c51771e38a5e88963425ae6 - that adds a new pure virtual method, so existing sub-classes of that need adjustments. Already worked around, so no fix required from our POV anymore.

Regarding the QR codes, I'll see if I can create a test case now that I know this is supposed to work. The failures I saw so far were unfortunately on non-public documents.

**Comment by axxel** [comment]
I see. My bad. Sorry for that.

But your fix is indeed non-optimal. Is there any chance I could convince you to completely do away with your `QImagePureBinarizer` and `MultiFormatReader` use and simply call into `ReadBarcode` like in the `ZXingQtReader` example? That API has been there since 1.0.8 (released in January). Or maybe do a big "if version >=1.1" and use the new API there and leave the old code (without your recent fix) for the else part and remove that from your code base as soon as you feel comfortable with.

Note: if you use `DecodeHints().setBinarizer(Binarizer::FixedThreshold).setIsPure(true)`, you get your `QImagePureBinarizer` functionality, only even faster. Your projects and that of @huycn are the only code bases that I know of that still rely on the `MultiFormatReader` + `LuminanceSource` API and I'd like to get rid of that to save even more copies and lower the code complexity further.

Also if you had some insight into how to properly provide the `ZXingQtReader.h` stuff as a header-only Qt wrapper as part of the 'official' API. E.g. any suggestions on how to make the `BarcodeFormat` enum visible to the MOC-system without repeating its definition? That would be greatly appreciated.

**Comment by vkrause** [comment]
That's actually what I tried first, but that showed the two issues I mentioned. That's why I went with the current indeed sub-optimal workaround for now, to unblock our nightly builds. I'll revisit that with your fix for the grayscale images and see if I can produce an isolated test case for what remains.

**Comment by axxel** [comment]
For me to be able to debug your QRCode issue, it would likely be sufficient if I had an image where the border and the finder patterns are still intact but the 'inside' could be deleted, so you would not risk sharing sensitive information but could still use one of the samples you already have.

**Comment by axxel** [comment]
@vkrause I assume you are busy with other more pressing issues again? And you will likely look at this the next time I break you build? No problem, if that is what it takes ;).

**Comment by vkrause** [comment]
@axxel you are right of course, sorry about that :)

I managed to create a synthetic test case that shows the problem now, https://invent.kde.org/pim/kitinerary/-/commit/155c9abcdb4de0a00064fc6518f6fd2a17f61b38 has a test file that can be freely used. Adding even just a 1px quiet zone to that on the top/left makes it work, otherwise this fails to be detected.

This is based on code images found in SNCF ticket PDFs, so nothing I can easily fix at the source to use a less weird image layout unfortunately.

**Comment by axxel** [comment]
Thanks for looking into this again. Looking at the sample immediately resolves the mystery why your QRCodes need a quite zone: they are Aztec codes :).

The Aztec detector has not been touched by me, yet, but I had plans to look into it for some time now. While I intend to completely rewrite the detector to be able to deal with rotated/skewed symbols, I'll have a quick look if I can fix your particular issue inside the current code base. I'll add your sample to our own black-box-tests.

**Comment by vkrause** [comment]
Aztec codes without any quiet zone do seem to work though: https://invent.kde.org/pim/kitinerary/-/blob/master/autotests/barcodes/aztec.png - so this problem does seem to be limited to hybrid cases?

(and yes, I indeed mixed up the code type when inspecting the failing tests back then, due to looking at the wrong document, sorry for the confusion).

**Comment by axxel** [comment]
Yes, the problem is the `WhiteRectDetector`, respectively the assumption that the Aztec code is (roughly) centered inside the image. I'm currently looking into the possibility to make it consider the pixels just outside the image to be 'white'. This should work and the only downside would be an increased runtime.

Question: will you call `ReadBarcode` with the `isPure` flag set to true for this use-case? That would open up another approach to solve this particular problem.

**Comment by vkrause** [comment]
`isPure` set to true seems to work for my entire test collection (apart from this problem here), so yes, enabling/requiring that is no problem.

**Comment by axxel** [comment]
@vkrause does my latest commit fix your remaining issues?

Also: to get the best performance, make sure to `setBinarizer(BoolCast)`. This will assume perfect black/white input. Slightly less efficient would be `FixedThreshold` which treats pixels as black if their luminance is `< 128`.

**Comment by vkrause** [comment]
Good news first: your fix works :)

Regarding BoolCast: This doesn't seem to work for us, due to parts of the input data not being "perfect" (JPEG artifacts, using slight color variations, etc). We could potentially track knowledge about the quality (e.g. for PDF vector graphics we know they are perfect), but the performance difference to FixedThreshold seems small enough for our use case to probably not make that worth it (a few 10s of milliseconds out of a total of about 10 seconds on a test run with about 1k documents).

So far so good, however during further testing I have found a relevant case for us that breaks in pure mode, see https://invent.kde.org/pim/kitinerary/-/commit/b2eab06fd95737278c019ddbf68465a1242f359e for a synthetic test case. The problem are 2D codes that contain individual rows/columns that are one pixel wider than they should be, probably due to scaling gone wrong somewhere. In non-pure mode this works fine. I have found cases "in the wild" at least for QR and DataMatrix, I haven't tested Aztec for this yet. Sorry for not having noticed that earlier, as that of course reverses my previous statement that unconditional pure mode is a viable option for us. Trying both modes is a possibility of course, I haven't measured yet how that performs compared to the old approach though.

**Comment by axxel** [comment]
`BoolCast`: I see how JPEGs would make this `BoolCast` approach break but are definitively a reasonable input. => `FixedThreshold` it is then.

`isPure`: your QRCode sample works now. I opted to make the `isPure` code path a little slower to facilitate this kind of input, because if you can't use it in your application, then it would be pretty useless altogether. Note: you can likely come up with a
deliberately constructed example of a symbol that fails again but works in the `!isPure` mode. Also, the DataMatrix code needs
to be updated as well. If you have an example for that, too, please let me know. The Aztec code should in fact work exactly the
same in this regards in both code paths.

**Comment by vkrause** [comment]
With the latest changes I get the following exception when encountering a large 910x910px Aztec code with a 10px dot size and no quiet zone. I haven't managed to create a synthetic test case for this yet, but I'll email you a somewhat censored version of a real code that triggers this as well.

#0  0x00007ffff6e94442 in __cxxabiv1::__cxa_throw (obj=obj@entry=0x5961f0, tinfo=0x7ffff6fb5228 <typeinfo for std::out_of_range>, dest=0x7ffff6ea8f10 <std::out_of_range::~out_of_range()>)
    at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:78
#1  0x00007ffff6e8b2ff in std::__throw_out_of_range_fmt (__fmt=__fmt@entry=0x7ffff5f02168 "vector::_M_range_check: __n (which is %zu) >= this->size() (which is %zu)") at ../../../../../libstdc++-v3/src/c++11/functexcept.cc:82
#2  0x00007ffff5e8c977 in std::vector<unsigned char, std::allocator<unsigned char> >::_M_range_check (this=0x44e8f0, this=0x44e8f0, __n=<optimized out>) at /usr/include/c++/10/bits/stl_algobase.h:1089
#3  std::vector<unsigned char, std::allocator<unsigned char> >::at (__n=<optimized out>, this=0x44e8f0) at /usr/include/c++/10/bits/stl_vector.h:1112
#4  ZXing::BitMatrix::get (i=<optimized out>, this=0x44e8e0) at /k/kde5/src/zxing-nubook/core/src/BitMatrix.h:69
#5  ZXing::BitMatrix::get (y=<optimized out>, x=<optimized out>, this=0x44e8e0) at /k/kde5/src/zxing-nubook/core/src/BitMatrix.h:128
#6  ZXing::BitMatrix::get (p=..., this=0x44e8e0) at /k/kde5/src/zxing-nubook/core/src/BitMatrix.h:280
#7  ZXing::BitMatrix::get (p=..., this=0x44e8e0) at /k/kde5/src/zxing-nubook/core/src/BitMatrix.h:281
#8  ZXing::Deflate (input=..., width=<optimized out>, height=<optimized out>, top=10, left=10, subSampling=20) at /k/kde5/src/zxing-nubook/core/src/BitMatrix.cpp:297
#9  0x00007ffff5ef8e47 in ZXing::QRCode::DetectPure (image=...) at /k/kde5/src/zxing-nubook/core/src/qrcode/QRDetector.cpp:310
#10 ZXing::QRCode::Detect (image=..., tryHarder=<optimized out>, isPure=<optimized out>) at /k/kde5/src/zxing-nubook/core/src/qrcode/QRDetector.cpp:321
#11 0x00007ffff5efb861 in ZXing::QRCode::Reader::decode (this=0x57a1b0, image=...) at /k/kde5/src/zxing-nubook/core/src/qrcode/QRReader.cpp:48
#12 0x00007ffff5e99221 in ZXing::MultiFormatReader::read (this=this@entry=0x7fffffffbe30, image=...) at /usr/include/c++/10/bits/unique_ptr.h:421
#13 0x00007ffff5e9b538 in ZXing::ReadBarcode (iv=..., hints=...) at /usr/include/c++/10/bits/shared_ptr_base.h:619
#14 0x00007ffff7db9275 in operator() (img=..., __closure=<synthetic pointer>) at /k/kde5/inst/include/ZXing/ReadBarcode.h:37
#15 KItinerary::BarcodeDecoder::decodeZxing (this=<optimized out>, img=..., format=..., result=...) at /data2/k/kde5/src/kitinerary/src/barcodedecoder.cpp:188

**Comment by axxel** [comment]
Thanks for alpha-testing this stuff. I should have seen this particular crash coming.

I'll further improve the usability of the `isPure` mode in the presence of aliasing for QRCodes (probably tomorrow).

**Comment by axxel** [comment]
I just improved the `QRCode::DetectPure()` function to a point where I claim it will work just as well as the non-isPure code path for all symbols as long as there is no skew or rotation and the image contains no black pixels that don't belong to the symbol. (That means the caveat about "possibility to construct an example that breaks it" from my earlier comment is removed.)

**Comment by vkrause** [comment]
I'm now hitting the following assert in the QR detector, with https://invent.kde.org/pim/kitinerary/-/blob/master/autotests/barcodes/aztec.png as input:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007ffff6af6864 in __GI_abort () at abort.c:79
#2  0x00007ffff6af6749 in __assert_fail_base (fmt=0x7ffff6c636e8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x7ffff5f11700 "cur.isBlack()", file=0x7ffff5f116c8 "/k/kde5/src/zxing-nubook/core/src/qrcode/QRDetector.cpp",
    line=178, function=<optimized out>) at assert.c:92
#3  0x00007ffff6b05f96 in __GI___assert_fail (assertion=assertion@entry=0x7ffff5f11700 "cur.isBlack()", file=file@entry=0x7ffff5f116c8 "/k/kde5/src/zxing-nubook/core/src/qrcode/QRDetector.cpp", line=line@entry=178,
    function=function@entry=0x7ffff5f11668 "double ZXing::QRCode::EstimateModuleSize(const ZXing::BitMatrix&, ZXing::PointF, ZXing::PointF)") at assert.c:101
#4  0x00007ffff5ef2e50 in ZXing::QRCode::EstimateModuleSize (image=..., a=..., b=...) at /k/kde5/src/zxing-nubook/core/src/BitMatrix.h:276
#5  0x00007ffff5ef303b in ZXing::QRCode::EstimateDimension (image=..., a=..., b=...) at /k/kde5/src/zxing-nubook/core/src/qrcode/QRDetector.cpp:195
#6  0x00007ffff5ef6b4e in ZXing::QRCode::DetectPure (image=...) at /k/kde5/src/zxing-nubook/core/src/Point.h:31
#7  ZXing::QRCode::Detect (image=..., tryHarder=<optimized out>, isPure=<optimized out>) at /k/kde5/src/zxing-nubook/core/src/qrcode/QRDetector.cpp:346
#8  0x00007ffff5ef9ee1 in ZXing::QRCode::Reader::decode (this=0x45fca0, image=...) at /k/kde5/src/zxing-nubook/core/src/qrcode/QRReader.cpp:48
#9  0x00007ffff5e97241 in ZXing::MultiFormatReader::read (this=this@entry=0x7fffffffbf30, image=...) at /usr/include/c++/10/bits/unique_ptr.h:421
#10 0x00007ffff5e99558 in ZXing::ReadBarcode (iv=..., hints=...) at /usr/include/c++/10/bits/shared_ptr_base.h:619
```

**Comment by axxel** [comment]
@vkrause Sorry for wasting your time with untested software but thanks for the feedback. I fixed the issue and added more tests.

I'm currently working a `pure` detector for PDF417 which will be able to detect upside-down and rotated symbols as well. That should allow you to remove your special casing for PDF417 codes altogether.

**Comment by vkrause** [comment]
Much better, down to one test failure now it seems, I'll send you the failing code by email, haven't managed to figure out what's special about that one yet.

**Comment by axxel** [comment]
Thanks for sending the (hopefully final) issue. The problem is that it is rotated by 90°. I assumed in 'pure' scenarios QRCodes would always be rendered in their default orientation. Since the symbols are always square there is no layout incentive to store them rotated. Is this a 'valid' sample that shows up in e-mails? If so, I'll have to make it work, I guess ;).

**Comment by vkrause** [comment]
Indeed, there is some image transformation involved there apparently. The PDF this comes from actually seems to show it horizontally flipped. Let me dig into how we extract this from the PDF, could very well be our code messing up the transformation somewhere.

**Comment by vkrause** [comment]
Yep, this turned out to be a vector graphics barcode and we didn't apply the transformation for that correctly when rasterizing it for passing it to ZXing. With that fixed it's detected without problems.

There's actually many more test documents affected by this, but all that went unnoticed so far as ZXing was handling some of that in non-pure mode apparently, and we have an explicit workaround for flipped PDF417 codes (which we had encountered in  raster barcodes in PDF files already before implementing vector barcode support). So this is slightly faster now since we will find a barcode on first try already, without having to try a flipped version :)

**Comment by axxel** [comment]
Very good to hear. So all your samples work with current HEAD in `isPure` mode?

Regarding the flipped PDF417, in case you have not noticed, thanks to @tailsu you don't need to deal with flipping images outside of ZXing anymore. My upcoming `pure` special casing for PDF417 will keep that feature and will additionally spare you from trying rotated symbols as well. With non-square symbols like PDF417 I see why people would choose to render them rotated onto their PDFs.

Anything else necessary for you to switch to `ReadBarcode()`?

**Comment by vkrause** [comment]
I've committed a code path using ReadBarcode now (and removed the usage of other deprecated API we still had in there while at it), conditional on a version number greater than 1.1.1, as we rely on all the recent fixes for this. So the next ZXing release would automatically complete the transition for us.

I'll revisit our hacks for rotations/flips, this predates a lot of improvements in ZXing as well as fixes for shortcomings in our PDF handling code, it is quite possible that parts or even all of that are no longer needed in this form.

**Comment by axxel** [comment]
Nice. A couple things I'd like to comment on:

* In the `ZXING_USE_READBARCODE` code path you only need to include `ZXing/ReadBarcode.h`
* The other two projects Qrca and Kaidan would need to be ported as well. Since they are only concerned about non-pure QRCodes and apparently only get Greyscale8 input, they can in fact use release 1.1.0 with its `ReadBarcode` implementation. Are you in a position to do that or do I need to contact someone else?
* Do you see any value in or have input for my proposed 'upstream' `ZXingQtReader.h` wrapper?

**Comment by axxel** [comment]
Addendum:
 * With my latest PDF417 improvements you should be able to drop all your rotation/flip special casing.
 * The new isPure approach should be considerably faster overall compared to the 1.1.1 code path. I'd be curious to get some feedback from you whether you can confirm that. I'd argue that your code that selects the decoders based on aspect ratio is not required anymore either.
 * Building for an architecture that support SSE or even AVX will make a difference, too.

**Comment by vkrause** [comment]
With the latest version I see a number of regressions with PDF417 decoding unfortunately, I'll email you a few samples.

Regarding the Qt wrapper, I don't see a way of providing Qt introspection features with a header-only setup. However the most valuable thing in there is the QImage -> ImageView adapter code IMHO, which should work header-only.

**Comment by Thaodan** [comment]
Talking about Qt integration: would it make sense to add a QML Plugin?

**Comment by axxel** [comment]
@Thaodan Have you seen [ZXingQtCamReader.cpp](https://github.com/nu-book/zxing-cpp/blob/master/example/ZXingQtCamReader.cpp)? Is there anything missing to make `ZXingQtReader.h` it a 'QML Plugin'?

**Comment by Thaodan** [comment]
> @Thaodan Have you seen
> [ZXingQtCamReader.cpp](https://github.com/nu-book/zxing-cpp/blob/master/exa
> mple/ZXingQtCamReader.cpp)? Is there anything missing to make
> `ZXingQtReader.h` it a 'QML Plugin'?
> 
No I don't but was thinking more of something that is in a form of library as 
a dynamic plugin so it can be shipped along the main library.
But this looks like a good start.


Also I might add QZXing already had better Qt integration in QImage and such.
This could be a good example to start up on in addition.

**Comment by axxel** [comment]
> Also I might add QZXing already had better Qt integration in QImage and such.

What do believe is missing?

**Comment by kosadev** [comment]
@axxel Do you think that it would be beneficial to put the QR scanner thing in the Qt repo? That would be definetely a good thing for the community, but I don't know how Qt maintainers will see that.

**Comment by axxel** [comment]
@kosadev I don't think I want to get the Qt company involved in this. My idea was rather to add a 'Qt wrapper' to this repository to prevent even more fragmentation: @swex started something similar to your effort 4 years ago but has seemingly abandoned it. Then there is a sizable community around https://github.com/ftylitak/qzxing. I got in touch with @ftylitak years ago about merging efforts but he was heavily invested in backward compatibility and did not want to commit to a c++14 build dependency. We are at c++17 now and I'm eyeing c++20 already, so there is likely not much point in bringing this up again (even though I am just doing it right now ;)).

The question is: do you see value in (partly) upstreaming your functionality? I did not have a close look at it, yet. It might be too specific to your own needs for me to consider it fitting. As you can see above and in the sample code, my idea was to have a very light header only wrapper that is trivial for Qt people to use. It got stuck there as an example because of lack of interest. I'm not against making it into a library/plugin if there is a technically convincing argument.

**Comment by vkrause** [comment]
Btw regarding generic Qt/ZXing integration, on the KDE side we meanwhile also have https://invent.kde.org/frameworks/prison/-/tree/master/src/scanner for scanning live video feeds from Qt Multimedia 5/6 (QML usage examples can be found further up in the same repo). That doesn't cover working with a single image though.

**Comment by vkrause** [comment]
Interesting indeed, I didn't know the SCodes implementation yet. One difference I can spot on a quick look is that this seems build around QString for the barcode content. This turned out to be problematic for us when dealing with binary rather than textual contents. That's why this is a bit more complex in our API, supporting both QString and QByteArray.

For barcode generation we actually have a dedicated library next to this as well, which is much older already and predates our use of ZXing (and thus uses other libraries and its own code to generate most formats). This is something I'd personally like to see phased out in favor of ZXing, but this isn't entirely up to me :)

Regarding the video frame formats, this should be the case for the formats we handle explicitly for zero-copy processing by ZXing, other formats that are either not fully specified or that cannot be directly consumed go through a QImage conversion first. That's provided by Qt, which hopefully knows how to deal with that specific format (it has to anyway to eventually render this on the screen).

**Comment by axxel** [comment]
> QString / QByteArray

Sure. You are not the first to have to deal with the issues arising during a `std::wstring` to native string object conversion when binary data is involved: @markusfisch had to come up with a hack in the android/JNI context as well. What I'm saying is: looks like a generic problem that belongs into such a wrapper in some shape or form.

> barcode generation

see new issue #332 for further discussion.

> video frame formats / row stride

I was not talking about the generation of `QImage` objects from `QVideoFrame` but the interpretation of `QVideoFrame` data in your zero-copy code. I strongly believe that you have a nasty bug there that will show up sooner or later, once someone gets there hands on some video-capture hardware that returns their frames with a row-stride other than `width x pixelSize` and I'm pretty sure they exist. That is the reason for the existence of the `bytesPerLine()` member.

> 'upstreaming' your code

Interestingly, you did not address my first question. I interpret that as a hint that the answer is closer to no than to yes. But I'd prefer to hear your opinion, instead of speculating about it ;).

**Comment by vkrause** [comment]
> video frame formats / row stride

Ah, I see what you mean. https://invent.kde.org/frameworks/prison/-/merge_requests/22 should address that.

> upstreaming

Sorry, not answering that part of the question was just me being sloppy, not in any way intentional :) I would indeed see value in a joined implementation over maintaining our own, as you say feature-wise this is all very close together anyway, so it's extremely likely that we'd be able to come up with something that matches our needs. Where exactly that is located is secondary IMHO, the main point for me is that it saves all of us some work.

Our current implementation is the result of unifying a few application-specific implementations we had of this, doing that inside our infrastructure was just the easiest and fastest way, and I admittedly didn't think about doing this in an even wider context.

**Comment by ApiTracer** [comment]
Hello,

I'm facing all issues raised by @axxel, so I want to contribute to the Qt6 port and the concentration of efforts to provide a great Qt port of zxing.
Some contexts of my use :
- I use zxing in my Android/IOS Qt app 
- The use is outdoor detection of small 1D code-128 barcodes. Some screenshots :
![scanner_screenshot](https://user-images.githubusercontent.com/74172673/171276846-5e010678-1e78-40b3-bc84-2949d211350d.png) (real screenshot)
![barcode_label](https://user-images.githubusercontent.com/74172673/171276512-c3a555a2-3d94-4cfe-8da5-d51b30892a71.png) (new label form, but not real image)
- Qt5 app and zxing integration works pretty well. I tested several integrations (zbar and severals zxing integrations), to finish with Scodes integration for the speed, in part due to this zxing-cpp. At qt5 time, goals was improve speed on old Android phone and iPhone, avoid lag after long scan period (typical use is few minutes active scanning to find several different barcodes : ~100), and decrease battery consumption.
- I needed to port my app to Qt6, in particular for billing integration of Android. Problems started with that.
- I start by trying to port Scodes to Qt6, then I try others integrations (Prison of KDE and others) and finish to find QZXing Qt6 branch. It works on IOS but slow and crash on Android. Thanks to this Qt-Bug(https://bugreports.qt.io/browse/QTBUG-97789), I can make works the QZXing on Android and IOS but it's too much slow compare to Qt5 Scodes, so I can't update my app.

I begin to try a Qt6 port of the Qt example of this repo, but it will be my first contribute to GitHub public project and I don't known well header-only. @axxel are you ok to help me to start the port in the right way and to help to make a general example if I success to make a working but sub-optimal project ?

**Comment by axxel** [comment]
@ApiTracer Thanks for joining the discussion. You wrote that you tried Prison of KDE but did not end up using it. May I ask why?

**Comment by ApiTracer** [comment]
The main reason, it's I haven't success to config all the KDE workspace, especially Extra-make-module. brew install of ECM works with qt creator after some settings but it raise errors. And it seems that it's not works with IOS.

**Comment by axxel** [comment]
I see. I would suggest to do the following: use the Prison code as an inspiration/template to upgrade the current `ZXingQtReader.h` for Qt6 support. If that works, we can then decide on how to move forward regarding the question of whether or not to make a Qt lib out of that.

**Comment by ApiTracer** [comment]
Hello,
I have a first try : https://github.com/ApiTracer/ZXingQtCamReader
It works on Android but it's very slow. The scanner make lags on UI thread.
On IOS, it's not works for now because the lib isn't linked in static. I have already fix it on another project, I will need to look it after. 

I tryed to avoid this lags, I think too much frames is scanned. But it crash when I add boolean or Timer to process only some frames. I think is linked to multithreading. 

Screenshot that show that the project works with the android simulator on macOS. but the TV of the android simulator show there are very few fps.
<img width="1002" alt="Capture d’écran 2022-06-02 à 15 19 51" src="https://user-images.githubusercontent.com/74172673/171655074-1bcfe926-cc20-4609-902a-fbf74030d018.png">

![Screenshot_1654175995](https://user-images.githubusercontent.com/74172673/171655437-088a1065-0d2d-42ea-9a9a-ecea67ed3420.png)

**Comment by axxel** [comment]
@ApiTracer thanks for sharing. I can't really evaluate the changes you made, as you copied the code into a new project, so I have no diff. If you want to pursue this further, please clone this repo and make your changes there. I saw that you simply modified the code to work with qt6. Please take another look at the KDE code and make similarly sure that it compiles with both qt5 and qt6.

**Comment by axxel** [comment]
@ApiTracer I just pushed a changeset porting the example code to qt6 while keeping the qt5 support. If you give it a try, please let me know how it worked out.

**Comment by ApiTracer** [comment]
Hello, Thanks for the work of porting example to qt6. I add some changes to focus on example project.
I will keep example inside zxing directory to facilitate review of changes. (https://github.com/ApiTracer/zxing-cpp)
The example is still very long to start camera and then the fps of the camera is very low.

**Comment by axxel** [comment]
Your performance problems are limited to android if I understood you correctly? The changes you made are limited to the `CMakeLists.txt` so I assume the qt6 port I did generally works?

I had a short look at what would be required for me to set up a qt-for-android build environment and it seems to me to be more involved than I am prepared to spend time on right now. So I'm afraid I can not help you with debugging your setup. [This comment](https://github.com/nu-book/zxing-cpp/issues/104#issuecomment-1155363925) contains some performance numbers of the native android app run on my Pixel 3. Maybe you can also give that sample app a try on your device to see what should be possible.

**Comment by ApiTracer** [comment]
Sorry for the late answer. The last time, I had compilation errors on IOS. Now, it compiles on IOS and Android but performances are very bad on both. With SCodes on qt5, I had real time perfs on most devices, but with this version with only code128 camera preview is very laggy with try harder, tryrotate and trydownscale on my iPhone 7.
It's quite ok when disable one or two and ok with three disabled but then barcode is very difficult to found while context is very good. Display time is around 120ms with only code 128 and try harder, tryrotate and try downscale activated. (And barcode is difficult to detect)Fork with example in the right directory : [](https://github.com/ApiTracer/zxing-cpp )

**Comment by axxel** [comment]
Wild guess: you are testing a debug build? I'm pretty sure there is something wrong with your build environment and not with the code itself.

**Comment by ApiTracer** [comment]
Yes, I test debug build. I try with release

**Comment by ApiTracer** [comment]
Thanks you very much. It's much better, sorry for your time ...

**Comment by ApiTracer** [comment]
Hi,
I come again because of performance issue. After my messages, I successfully used your example as base to make a wrapper used in my application in Qt6 with zxing-cpp 1.4.0 (8 July 2022).
Scanner is really fast on most devices but some Android device are laggy. 
So I try to update to master branch of zxing-cpp but the update created lag in live camera display of the app. (Without any changes of wrapper)
I think is linked to multithreading but I don't understand why it can changes with zxing-cpp submodule update.
I look to each commits created between the commit used before and the master current state but I not found changes of function that can explain the changes.

Can you help me to debug the changes ? 

Fork of zxing-cpp, master is branch after update and before, I use commit wit this SHA : 9c1e84a9837b7a0b4b44504b9437eb464a93657b
[https://github.com/ApiTracer/zxing-cpp/tree/master](url)

Screen recording before update (iPhone 7) :
https://user-images.githubusercontent.com/74172673/217846490-8e3c8be5-fe8e-43aa-8b1c-ae86987c7831.MP4

Screen recording after update (iPhone 7) :
https://user-images.githubusercontent.com/74172673/217846564-748426a1-4ae7-4ad6-889d-8b543bee4a95.MP4

Customized Wrapper (changes are made to avoid laggy UI and to decrease CPU impact for recent mobile and to add QrCode generated image accessible from QML) :
`/*
 * Copyright 2020 Axel Waggershauser
 */
// SPDX-License-Identifier: Apache-2.0

#pragma once

#include "ReadBarcode.h"
#include "BarcodeFormat.h"
#include "BitMatrix.h"
#include "BitMatrixIO.h"
#include "CharacterSet.h"
#include "MultiFormatWriter.h"
#include "TextUtfEncoding.h"

#include <QImage>
#include <QDebug>
#include <QMetaType>
#include <QScopeGuard>

#ifdef QT_MULTIMEDIA_LIB
#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
#include <QAbstractVideoFilter>
#else
#include <QVideoFrame>
#include <QVideoSink>
#endif
#include <QElapsedTimer>
#include <QtConcurrent>
#include <QRect>
#include <QQuickImageProvider>
#include <QImage>
#endif

// This is some sample code to start a discussion about how a minimal and header-only Qt wrapper/helper could look like.

namespace ZXingQt {

Q_NAMESPACE

//TODO: find a better way to export these enums to QML than to duplicate their definition
// #ifdef Q_MOC_RUN produces meta information in the moc output but it does end up working in qml
#ifdef QT_QML_LIB
enum class BarcodeFormat
{
    None            = 0,         ///< Used as a return value if no valid barcode has been detected
    Aztec           = (1 << 0),  ///< Aztec
    Codabar         = (1 << 1),  ///< Codabar
    Code39          = (1 << 2),  ///< Code39
    Code93          = (1 << 3),  ///< Code93
    Code128         = (1 << 4),  ///< Code128
    DataBar         = (1 << 5),  ///< GS1 DataBar, formerly known as RSS 14
    DataBarExpanded = (1 << 6),  ///< GS1 DataBar Expanded, formerly known as RSS EXPANDED
    DataMatrix      = (1 << 7),  ///< DataMatrix
    EAN8            = (1 << 8),  ///< EAN-8
    EAN13           = (1 << 9),  ///< EAN-13
    ITF             = (1 << 10), ///< ITF (Interleaved Two of Five)
    MaxiCode        = (1 << 11), ///< MaxiCode
    PDF417          = (1 << 12), ///< PDF417 or
    QRCode          = (1 << 13), ///< QR Code
    UPCA            = (1 << 14), ///< UPC-A
    UPCE            = (1 << 15), ///< UPC-E
    MicroQRCode     = (1 << 16), ///< Micro QR Code

    LinearCodes = Codabar | Code39 | Code93 | Code128 | EAN8 | EAN13 | ITF | DataBar | DataBarExpanded | UPCA | UPCE,
    MatrixCodes = Aztec | DataMatrix | MaxiCode | PDF417 | QRCode | MicroQRCode,
};

enum class ContentType { Text, Binary, Mixed, GS1, ISO15434, UnknownECI };

#else
using ZXing::BarcodeFormat;
using ZXing::ContentType;
#endif

using ZXing::DecodeHints;
using ZXing::Binarizer;
using ZXing::BarcodeFormats;

Q_ENUM_NS(BarcodeFormat)
Q_ENUM_NS(ContentType)

template<typename T, typename = decltype(ZXing::ToString(T()))>
QDebug operator<<(QDebug dbg, const T& v)
{
    return dbg.noquote() << QString::fromStdString(ToString(v));
}

class Position : public ZXing::Quadrilateral<QPoint>
{
    Q_GADGET

    Q_PROPERTY(QPoint topLeft READ topLeft)
    Q_PROPERTY(QPoint topRight READ topRight)
    Q_PROPERTY(QPoint bottomRight READ bottomRight)
    Q_PROPERTY(QPoint bottomLeft READ bottomLeft)

    using Base = ZXing::Quadrilateral<QPoint>;

public:
    using Base::Base;
};

class Result : private ZXing::Result
{
    Q_GADGET

    Q_PROPERTY(BarcodeFormat format READ format)
    Q_PROPERTY(QString formatName READ formatName)
    Q_PROPERTY(QString text READ text)
    Q_PROPERTY(QByteArray bytes READ bytes)
    Q_PROPERTY(bool isValid READ isValid)
    Q_PROPERTY(ContentType contentType READ contentType)
    Q_PROPERTY(Position position READ position)

    QString _text;
    QByteArray _bytes;
    Position _position;

public:
    Result() = default; // required for qmetatype machinery

    explicit Result(ZXing::Result&& r) : ZXing::Result(std::move(r)) {
        _text = QString::fromStdString(ZXing::Result::text());
        _bytes = QByteArray(reinterpret_cast<const char*>(ZXing::Result::bytes().data()), Size(ZXing::Result::bytes()));
        auto& pos = ZXing::Result::position();
        auto qp = [&pos](int i) { return QPoint(pos[i].x, pos[i].y); };
        _position = {qp(0), qp(1), qp(2), qp(3)};
    }

    using ZXing::Result::isValid;

    BarcodeFormat format() const { return static_cast<BarcodeFormat>(ZXing::Result::format()); }
    ContentType contentType() const { return static_cast<ContentType>(ZXing::Result::contentType()); }
    QString formatName() const { return QString::fromStdString(ZXing::ToString(ZXing::Result::format())); }
    const QString& text() const { return _text; }
    const QByteArray& bytes() const { return _bytes; }
    const Position& position() const { return _position; }

    // For debugging/development
    int runTime = 0;
    Q_PROPERTY(int runTime MEMBER runTime)
};

inline QList<Result> QListResults(ZXing::Results&& zxres)
{
    QList<Result> res;
    for (auto&& r : zxres)
        res.push_back(Result(std::move(r)));
    return res;
}

inline QList<Result> ReadBarcodes(const QImage& img, const DecodeHints& hints = {})
{
    using namespace ZXing;

    auto ImgFmtFromQImg = [](const QImage& img) {
        switch (img.format()) {
        case QImage::Format_ARGB32:
        case QImage::Format_RGB32:
#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
            return ImageFormat::BGRX;
#else
            return ImageFormat::XRGB;
#endif
        case QImage::Format_RGB888: return ImageFormat::RGB;
        case QImage::Format_RGBX8888:
        case QImage::Format_RGBA8888: return ImageFormat::RGBX;
        case QImage::Format_Grayscale8: return ImageFormat::Lum;
        default: return ImageFormat::None;
        }
    };

    auto exec = [&](const QImage& img) {
        return QListResults(ZXing::ReadBarcodes(
            {img.bits(), img.width(), img.height(), ImgFmtFromQImg(img), static_cast<int>(img.bytesPerLine())}, hints));
    };

    return ImgFmtFromQImg(img) == ImageFormat::None ? exec(img.convertToFormat(QImage::Format_Grayscale8)) : exec(img);
}

inline Result ReadBarcode(const QImage& img, const DecodeHints& hints = {})
{
    auto res = ReadBarcodes(img, DecodeHints(hints).setMaxNumberOfSymbols(1));
    return !res.isEmpty() ? res.takeFirst() : Result();
}

#ifdef QT_MULTIMEDIA_LIB
inline QList<Result> ReadBarcodes(const QVideoFrame& frame, const DecodeHints& hints = {}, const QRect captureRect = QRect())
{
    using namespace ZXing;

    auto img = frame; // shallow copy just get access to non-const map() function
#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
    if (!frame.isValid() || !img.map(QAbstractVideoBuffer::ReadOnly)){
#else
    if (!frame.isValid() || !img.map(QVideoFrame::ReadOnly)){
#endif
        qWarning() << "invalid QVideoFrame: could not map memory";
        return {};
    }
    auto unmap = qScopeGuard([&] { img.unmap(); });

    ImageFormat fmt = ImageFormat::None;
    int pixStride = 0;
    int pixOffset = 0;

#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
#define FORMAT(F5, F6) QVideoFrame::Format_##F5
#define FIRST_PLANE
#else
#define FORMAT(F5, F6) QVideoFrameFormat::Format_##F6
#define FIRST_PLANE 0
#endif

    switch (img.pixelFormat()) {
    case FORMAT(ARGB32, ARGB8888):
    case FORMAT(ARGB32_Premultiplied, ARGB8888_Premultiplied):
    case FORMAT(RGB32, RGBX8888):
#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
        fmt = ImageFormat::BGRX;
#else
        fmt = ImageFormat::XRGB;
#endif
        break;

    case FORMAT(BGRA32, BGRA8888):
    case FORMAT(BGRA32_Premultiplied, BGRA8888_Premultiplied):
    case FORMAT(BGR32, BGRX8888):
#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
        fmt = ImageFormat::RGBX;
#else
        fmt = ImageFormat::XBGR;
#endif
        break;

#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
    case QVideoFrame::Format_RGB24: fmt = ImageFormat::RGB; break;
    case QVideoFrame::Format_BGR24: fmt = ImageFormat::BGR; break;
    case QVideoFrame::Format_YUV444: fmt = ImageFormat::Lum, pixStride = 3; break;
#else
    case QVideoFrameFormat::Format_P010:
    case QVideoFrameFormat::Format_P016: fmt = ImageFormat::Lum, pixStride = 1; break;
#endif

    case FORMAT(AYUV444, AYUV):
    case FORMAT(AYUV444_Premultiplied, AYUV_Premultiplied):
#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
        fmt = ImageFormat::Lum, pixStride = 4, pixOffset = 3;
#else
        fmt = ImageFormat::Lum, pixStride = 4, pixOffset = 2;
#endif
        break;

    case FORMAT(YUV420P, YUV420P):
    case FORMAT(NV12, NV12):
    case FORMAT(NV21, NV21):
    case FORMAT(IMC1, IMC1):
    case FORMAT(IMC2, IMC2):
    case FORMAT(IMC3, IMC3):
    case FORMAT(IMC4, IMC4):
    case FORMAT(YV12, YV12): fmt = ImageFormat::Lum; break;
    case FORMAT(UYVY, UYVY): fmt = ImageFormat::Lum, pixStride = 2, pixOffset = 1; break;
    case FORMAT(YUYV, YUYV): fmt = ImageFormat::Lum, pixStride = 2; break;

    case FORMAT(Y8, Y8): fmt = ImageFormat::Lum; break;
    case FORMAT(Y16, Y16): fmt = ImageFormat::Lum, pixStride = 2, pixOffset = 1; break;

#if (QT_VERSION >= QT_VERSION_CHECK(5, 13, 0))
    case FORMAT(ABGR32, ABGR8888):
#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
        fmt = ImageFormat::RGBX;
#else
        fmt = ImageFormat::XBGR;
#endif
        break;
#endif
#if (QT_VERSION >= QT_VERSION_CHECK(5, 14, 0))
    case FORMAT(YUV422P, YUV422P): fmt = ImageFormat::Lum; break;
#endif
    default: break;
    }

    if (fmt != ImageFormat::None) {
        return QListResults(ZXing::ReadBarcodes(
                                ImageView(img.bits(FIRST_PLANE) + pixOffset, img.width(), img.height(), fmt, img.bytesPerLine(FIRST_PLANE), pixStride).cropped(captureRect.left(), captureRect.top(), captureRect.width(), captureRect.height()), hints));
    } else {
#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
        if (QVideoFrame::imageFormatFromPixelFormat(img.pixelFormat()) != QImage::Format_Invalid)
            return ReadBarcodes(img.image(), hints);
        qWarning() << "unsupported QVideoFrame::pixelFormat";
        return {};
#else
        return ReadBarcodes(img.toImage().copy(captureRect), hints);
#endif
    }
}

inline Result ReadBarcode(const QVideoFrame& frame, const DecodeHints& hints = {}, const QRect captureRect = QRect())
{
    auto res = ReadBarcodes(frame, DecodeHints(hints).setMaxNumberOfSymbols(1), captureRect);
    return !res.isEmpty() ? res.takeFirst() : Result();
}

#define ZQ_PROPERTY(Type, name, setter) \
public: \
    Q_PROPERTY(Type name READ name WRITE setter NOTIFY name##Changed) \
    Type name() const noexcept { return DecodeHints::name(); } \
    Q_SLOT void setter(const Type& newVal) \
    { \
        if (name() != newVal) { \
            DecodeHints::setter(newVal); \
            emit name##Changed(); \
        } \
    } \
    Q_SIGNAL void name##Changed();



#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
class BarcodeReader : public QAbstractVideoFilter, private DecodeHints
#else
class BarcodeReader : public QObject, private DecodeHints
#endif
{
    Q_OBJECT

public:
#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
    BarcodeReader(QObject* parent = nullptr) : QAbstractVideoFilter(parent) {}
#else
    BarcodeReader(QObject* parent = nullptr) : QObject(parent), m_decoding(false), m_active(true) {
    }

    ~BarcodeReader(){
        if (_sink)
            disconnect(_sink, nullptr, this, nullptr);
    }
#endif

    // TODO: find out how to properly expose QFlags to QML
    // simply using ZQ_PROPERTY(BarcodeFormats, formats, setFormats)
    // results in the runtime error "can't assign int to formats"
    Q_PROPERTY(int formats READ formats WRITE setFormats NOTIFY formatsChanged)

    Q_PROPERTY(QRect captureRect READ captureRect WRITE setCaptureRect NOTIFY captureRectChanged)

    int formats() const noexcept
    {
        auto fmts = DecodeHints::formats();
        return *reinterpret_cast<int*>(&fmts);
    }
    Q_SLOT void setFormats(int newVal)
    {
        if (formats() != newVal) {
            DecodeHints::setFormats(static_cast<ZXing::BarcodeFormat>(newVal));
            emit formatsChanged();
            qDebug() << DecodeHints::formats();
        }
    }
    Q_SIGNAL void formatsChanged();

    QRect captureRect() const{
        return m_captureRect;
    }
    Q_SLOT void setCaptureRect(const QRect &captureRect){
        if (captureRect == m_captureRect) {
            return;
        }
        m_captureRect = captureRect;
        emit captureRectChanged();
    }
    Q_SIGNAL void captureRectChanged();

    ZQ_PROPERTY(bool, tryRotate, setTryRotate)
    ZQ_PROPERTY(bool, tryHarder, setTryHarder)
    ZQ_PROPERTY(bool, tryDownscale, setTryDownscale)

public slots:
    ZXingQt::Result process(const QVideoFrame& frame)
    {
        if(m_active && !m_decoding && (!m_processTimer.isValid() || m_processTimer.elapsed() >= 10)){
            m_decoding = true;
            qDebug() << "Decoding : Time between last decode : " << m_processTimer.elapsed();
#ifdef Q_OS_ANDROID
            bool didWeMapTheFrame = false;
            if(!static_cast<QVideoFrame>(frame).isMapped()){
                _sink->setRhi(nullptr); // https://bugreports.qt.io/browse/QTBUG-97789
                didWeMapTheFrame = static_cast<QVideoFrame>(frame).map(QVideoFrame::ReadOnly);
            }

            if(static_cast<QVideoFrame>(frame).isMapped()){

                auto img = static_cast<QVideoFrame>(frame).toImage();

                if(!img.isNull()){
                    auto ret = QtConcurrent::run([=](){
                        QElapsedTimer t;
                        t.start();

                        auto res = ReadBarcode(img.copy(m_captureRect), *this);

                        res.runTime = t.elapsed();
                        if (res.isValid() && m_active)
                            emit foundBarcode(res);
                        m_decoding = false;
                        m_processTimer.restart();
                        return res;
                    });
                }else{
                    m_decoding = false;
                    m_processTimer.restart();
                }

                if(didWeMapTheFrame) static_cast<QVideoFrame>(frame).unmap();
                return ZXingQt::Result();
            }
            m_decoding = false;
            m_processTimer.restart();
            return ZXingQt::Result();
#else
            auto res = ReadBarcode(frame, *this, m_captureRect);
            if (res.isValid() && m_active)
                emit foundBarcode(res);
            m_decoding = false;
            m_processTimer.restart();
            return res;
#endif
        }
        else{
            return ZXingQt::Result();
        }
    }

signals:
    void foundBarcode(ZXingQt::Result result);

#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
public:
    QVideoFilterRunnable *createFilterRunnable() override;
#else
private:
    QVideoSink *_sink = nullptr;
    QElapsedTimer m_processTimer;
    bool m_decoding;
    bool m_active;

    /*!
    * \brief Capture area
    */
    QRect m_captureRect;

public:
    void setVideoSink(QVideoSink* sink) {
        if (_sink == sink)
            return;

        if (_sink)
            disconnect(_sink, nullptr, this, nullptr);

        _sink = sink;
        connect(_sink, &QVideoSink::videoFrameChanged, this, &BarcodeReader::process
    #ifdef Q_OS_ANDROID
                , Qt::DirectConnection
    #endif
                );
        m_active = true;
    }
    Q_PROPERTY(QVideoSink* videoSink WRITE setVideoSink)
    Q_INVOKABLE void stopFilter(){
        qDebug("disconnect videoFrameChanged");
        m_active = false;
        disconnect(_sink, nullptr, this, nullptr);
    }
#endif

};

#undef ZX_PROPERTY

#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
class VideoFilterRunnable : public QVideoFilterRunnable
{
    BarcodeReader* _filter = nullptr;

public:
    explicit VideoFilterRunnable(BarcodeReader* filter) : _filter(filter) {}

    QVideoFrame run(QVideoFrame* input, const QVideoSurfaceFormat& /*surfaceFormat*/, RunFlags /*flags*/) override
    {
        _filter->process(*input);
        return *input;
    }
};

inline QVideoFilterRunnable* BarcodeReader::createFilterRunnable()
{
    return new VideoFilterRunnable(this);
}
#endif

class ZXingImageProvider : public QQuickImageProvider
{
public:
    ZXingImageProvider() : QQuickImageProvider(QQuickImageProvider::Image){

    }

    QImage requestImage(const QString &id, QSize *size, const QSize &requestedSize){
        QImage image;
        if(id != ""){

            int slashIndex = id.indexOf('/');
            if (slashIndex == -1)
            {
                qWarning() << "Can't parse url" << id << ". Usage is encode/<data>";
                return QImage();
            }

            //Detect operation (ex. encode)
            QString operationName = id.left(slashIndex);
            if(operationName != "encode")
            {
                qWarning() << "Operation not supported: " << operationName;
                return QImage();
            }

            QString data;

            data = id.mid(slashIndex + 1);

            if(data != ""){

                ZXing::BarcodeFormat format = ZXing::BarcodeFormat::QRCode;

                int width = requestedSize.width(), height = requestedSize.height();

                int margin = 10;
                int eccLevel = 1;
                auto writer = ZXing::MultiFormatWriter(format).setMargin(margin).setEccLevel(eccLevel);
                auto matrix = writer.encode(data.toStdString(), width, height);
                image = QImage(width, height, QImage::Format_ARGB32);

                image.fill(qRgba(255, 255, 255, 255));
                for (uint i=0; i<width; ++i) {
                    for (uint j=0; j<height; ++j) {
                        if(i<matrix.width() && j<matrix.height()){
                            if (matrix.get(j,i)){
                                image.setPixel(i, j, qRgba(0, 0, 0, 255));
                            }
                        }
                    }
                }
            }
        }
        return image;
    }
};

#endif // QT_MULTIMEDIA_LIB

} // namespace ZXingQt


Q_DECLARE_METATYPE(ZXingQt::Position)
Q_DECLARE_METATYPE(ZXingQt::Result)

#ifdef QT_QML_LIB

#include <QQmlEngine>

namespace ZXingQt {

inline void registerQmlAndMetaTypes(QQmlEngine& engine)
{
    qRegisterMetaType<ZXingQt::BarcodeFormat>("BarcodeFormat");
    qRegisterMetaType<ZXingQt::ContentType>("ContentType");

    // supposedly the Q_DECLARE_METATYPE should be used with the overload without a custom name
    // but then the qml side complains about "unregistered type"
    qRegisterMetaType<ZXingQt::Position>("Position");
    qRegisterMetaType<ZXingQt::Result>("Result");

    qmlRegisterUncreatableMetaObject(
        ZXingQt::staticMetaObject, "ZXing", 1, 0, "ZXing", "Access to enums & flags only");
    qmlRegisterType<ZXingQt::BarcodeReader>("ZXing", 1, 0, "BarcodeReader");

    //    ZXingImageProvider provider;
    engine.addImageProvider("ZXing", new ZXingImageProvider());
}

} // namespace ZXingQt

#endif // QT_QML_LIB
`

QML use : 
`    BarcodeReader {
        id: barcodeReader

        videoSink: videoOutput.videoSink

        formats: ZXing.Code128
        tryRotate: true
        tryHarder: true
        tryDownscale: true

        captureRect: Qt.rect(videoOutput.sourceRect.width / 10, 0.3*videoOutput.sourceRect.height , videoOutput.sourceRect.width * (8/10), 0.4*videoOutput.sourceRect.height)

        // callback with parameter 'result', called for every successfully processed frame
         onFoundBarcode: function(result) {
            if(result.isValid && result.text != ""){
                if(scanningType == 0){
                    if(!busy && scanActive){
                        busy = true;
                        updateUserPos();
                        inputDataModel.newScan(scanningType, result.text, apiaryId, userCoordinate);
                        console.log("New barcode : ", result.text);
                    }
                }
                else{
                    updateUserPos();
                    inputDataModel.newScan(scanningType, result.text, apiaryId, userCoordinate);
                    console.log("New barcode : ", result.text);
                }
            }
        }
    }`

**Comment by axxel** [comment]
Sorry but I don't have the capacity to debug the performance regression in your code. If you suspect a change in the library between 1.4 and 2.0 is to blame, then I suggest you perform a bisection of the changes to find out which commit exactly caused the regression for your. Then I can likely help in fixing the issue.

**Comment by ApiTracer** [comment]
Hi,
I finally found the commit creating the issue : 983b3ad187cd2a6ee2bc33b9277e61a80df72eb3

Revert it (from GitHub Desktop) allows to retrieve normal behavior.

**Comment by axxel** [comment]
Thanks for the feedback. I looked real hard at that 2-line diff and can't find any reason why this would impact your performance, because I believe the code that gets executed in your use case should be exactly the same as without the change.

Can you maybe add some debugging code to your setup to see if either the `getBlackMatrix()` or the `flippAll()` function gets executed more frequently with the 'bad' change compared to without it?

How come you suspect that multithreading has anything to do with it? Are you somewhere actively using that?

**Comment by ApiTracer** [comment]
Multithreading is maybe more a wish than the problem : I think I have to make scanner run in another thread than UI thread to avoid micro-freeze of live feed display. 
Even with normal behavior, there is a little freeze for each decoded frame and for some devices the balance between UI and scanner work is broken and live feed is very laggy or live feed is correct but the scanner don't work as expected (barcode decoded sometimes but unusable due to the lake of reactivity)

Have you a easy way to add counter for these fonctions calls ?

**Comment by ApiTracer** [comment]
I was looking for something too complicated, I added some printf in each getBlackMatrix() (BinaryBitmap, HybridBinarizer and GlobalHistogramBinarizer) and flipAll() function.
With the commit, there is one getBlackMatrix() call in BinaryBitmap, one in HybridBinarizer and one call of flipAll. 
No call to these functions without the commit.

**Comment by axxel** [comment]
> With the commit, there is one getBlackMatrix() call in BinaryBitmap, one in HybridBinarizer and one call of flipAll.
> No call to these functions without the commit.

If there was no call to `getBlackMatrix()` at all, then you would not get any results back, ever, except you were only looking for linear (1D) symbologies? If you did not set the `DecodeHints::setFormats()` or the format contains any 2D code, then there must be at least 1 `getBlackMatrix` call for each `ReadBarcodes()` call. There may be more if your input is large enough and you have `tryDownscale` enabled (which is the default).

**Comment by ApiTracer** [comment]
Yes, sorry for the lake of precision. I search only code128 linear barcodes (and with always 6 digits)

**Comment by axxel** [comment]
Ahh, now I understand the situation. Requiring me to think about the relevance of the selected formats in this context with your confirmation that you are only looking for 1D codes made it click. The commit made you pay for stuff that you don't need. I will revert it. Thanks for the help in debugging this.

**Comment by ApiTracer** [comment]
Thanks you for your time and quick answers

**Comment by emericg** [comment]
I have made a demo application for Qt6 and barcode scanning.

https://github.com/emericg/QmlMobileScanner

Since Qt6 and the new Qt Multimedia module, the camera feature (and everything else unfortunately) is mostly broken, so success will depend a lot on what version is used (tested successfully with 6.5.3 and 6.6.1) and also what backend is chosen at run time (the new new ffmpeg backend for Qt Multimedia is at this time even more broken than the old backends).
This app forces the old android backend for instance, because the ffmpeg one cannot pass frames through video filters anymore.
There are also various Qt crashes when you stop cameras, or minimize apps and bring them back into focus... Hopefully everything will be squared away with time.

- The "official" qzxing project is mostly broken on Qt6, so the backend included in QmlMobileScanner has multiple fixes and improvements, and now run well with Android and iOS.

- The "official" zxingcpp Qt demo is a bit limited, so I've made a proper backend for the needs of this demo application, which adds efficient scanning using QtConcurrent and an image provider to generate barcodes with ease.
It's still about 2 to 3 times slower than qzxing, but it can scan a lot more formats obviously!

- The zint backend is also new, also with an image provider to generate very exotic barcodes with simple integration into Qml applications. It's a toy more than a proper backend, but it is quite fun to generate GridMatrix or UltraCodes :)
